### ISSUE TITLE
Have decorator `check_quantity` enforce specified unit

### ISSUE THREAD

**Issue by rocco8773** [PlasmaPy/PlasmaPy#620]
Currently [`check_quantity`](https://github.com/PlasmaPy/PlasmaPy/blob/efd7bff5bf34c64e70259886c9b4111b046f4d79/plasmapy/utils/checks.py#L20) does not enforce the specified unit. For example,
```
>>> import astropy.units as u
>>> from plasmapy.utils import check_quantity
>>>
>>> @check_quantity(**{'var': {'units': u.cm}}
... def show_unit(var: u.Quantity):
...     print(var)
>>> show_unit(5. * u.km)
5.0 km
```

I propose that `check_quantity` should enforce the units, such that...
```
>>> show_unit(5. * u.km)
500000.0 cm
```
I'm assuming the unit(s) specified to `check_quantity` are the one(s) you'll want to use in the decorated function.  If `check_quantity` is already examining the units, I don't see why it should not return the units you'll want to used.

For the case that multiple units are specified, then said units must not be equivalent.  That is, `var` would only be convertible to one of the specified units.

If this is a desired/accepted functionality, then I can initiate a PR.  I've already done this for a [branch on `bapsflib`](https://github.com/BaPSF/bapsflib/blob/add_astropy_units_to_plasma/bapsflib/utils/helpers.py), along with including an `equivalencies` key in the validations dictionary.  It does need a little cleaning and additional tests before establishing a PR.

**Comment by StanczakDominik** [comment]
Huh. If I understand it correctly, as it stands, `check_quantity` will 

1. error for your `show_unit` function if I were to input something unrelated to distance such as `5 * u.kg`
2. pass silently when putting in `5 * u.m` or any unit of distance as the argument, assuming that `astropy.units` can handle any necessary conversions
3. do nothing when putting in `5 * u.cm` as specified (which is kind of a subset of 2)

However, I'm not sure this is all that wrong if we are to trust `astropy.units`. Whenever we use the quantity values directly in functions, we tend to use `var.si.value` anyway. Consider a slight rewrite:

```
>>> import astropy.units as u
>>> from plasmapy.utils import check_quantity
>>>
>>> @check_quantity(**{'var': {'units': u.cm}}
... def show_unit(var: u.Quantity) -> u.cm:     # annotate return type here
...     return var                              # return instead of print
>>> show_unit(5. * u.km)
500000.0 cm
```

This is either something we already have, something `astropy.units.quantity_input` has or something we wanted to have.

I'm not sure it's necessary to hard convert the units on input if care is taken to get the quantity values via `si`. I'd be happy to discuss this point further, though!

@PlasmaPy/plasmapy-community ?

**Comment by rocco8773** [comment]
Yes, your understanding of how `check_quantity` currently works is correct.  It also looks like `astropy.units.quantity_input` just checks unit equivalencies and does not actually convert the passed argument.

My view on why to convert:
1. As a new user, my initial view on how `check_quantity` worked was that it enforced the units.  This is what the docstrings indicate.  If the behavior of `check_quantity` does not change, then I suggest changing the docstring wording to indicate **"equivalent units"**.
1. Not converting to the expected units leads to extra lines of code. `_check_quantity` is already doing `arg.units.to()` to check unit convertibility/equivalence.  Why not go the extra step, after establishing convertibility, to convert the actual argument `arg.to()`?  This conversion is already built into `astropy` and we tend to use `astropy` in the decorated function to convert.
1. For the handful of cases where the programmer wants to do the conversion (him)herself, then add a `'enforce'` key to the validations dictionary that is by default `True`. Or, one could build a new `astropy` equivalencies for a custom conversion (e.g. my [temperature_and_energy()](https://github.com/BaPSF/bapsflib/blob/20f86ce67d5403e4bb8b62393c2a1b48b55e5e79/bapsflib/utils/helpers.py#L30)), in conjunction with adding an `'equivalencies'` key to the validation dictionary.
1. This change should not affect any of the currently decorated functions.

**Comment by namurphy** [comment]
This is a really worthwhile discussion.  I'm still thinking about it, but my initial thoughts are:

- Having `check_quantity` convert the returned value to the return annotation is a really worthwhile idea.  If we do that, we should probably rename `check_quantity` to indicate that it is doing more than just checking the quantity.
- For the most part I'm indifferent/undecided on converting the units of the positional and keyword arguments as they are passed in, since ideally this should be taken care of by converting the return annotation.  The exception is eV, which should be converted to an actual temperature, and `check_quantity` is probably the best place to do it.
- In terms of process, this seems like a good opportunity for test-driven development.  Writing the tests first would help us figure out the behavior that we want and allow for some more discussion.  We could then iterate and get feedback on the tests before making changes to the code.
- I like the idea for the `'enforce'` keyword.
- There is definitely room for improvement in `check_quantity`!  I'm thinking in particular to some of my own early contributions to it from back when I knew less Python.
- I have to go to an arduous meeting now but I'd rather keep talking about this.

**Comment by rocco8773** [comment]
My thoughts:

* If the `'enforce'` key is added then I think the `check_quantity` name is still sufficient.
* Converting on the return notation is only useful if you are working in SI.  This becomes problematic when you're working with a Gaussian cgs equation, or another naturalized unit equaiton.  In some situations `astropy.units.Quantity.to()` just does not know what to do.  I know, within the function you could still work with the SI equation and then convert on the return, but...ugh.  However, by incorporating the `'enforce'` key either situation could be covered and the programmer could choose how he or she would want to handle the conversion.

I've already made most of this functionality for a `bapsflib` version.  If there's no objections, then I'll write some tests for it and open a PR to facilitate further discussion.  Now I just got to learn how to use `pytest`. 🤔

**Comment by StanczakDominik** [comment]
All right, I think I'm convinced now. Fully agree, let's do this!

Pytest is pretty handy and the existing tests should be either a good starting point or (hopefully not too much of) a mess. Feel free to reach out if you need a hand!

**Comment by namurphy** [comment]
Sounds good to me too!

**Comment by namurphy** [comment]
> Now I just got to learn how to use `pytest`. :thinking:

I forgot to mention that our [development guide](http://docs.plasmapy.org/en/latest/development/index.html) has a section on [testing](http://docs.plasmapy.org/en/latest/development/testing_guide.html).  It's been updated a lot since our last release so it's best to go to `latest` rather than `stable`.  We're still in the process of writing our dev guide, so please let us know if there are any parts that could benefit from more updates.  Thanks!

**Comment by rocco8773** [comment]
Thanks @namurphy, I'll keep those in mind and let you know if I have any comments.  I'm just much more familiar with `unittests`, so hopefully the learning curve for `pytest` won't be that bad.  ...I've actually already written the tests with `unittests` and now it's a matter of translating them.

Is there a preference for `enforce=True` to be default, or `False`? I tend to lean towards `True`.

**Comment by namurphy** [comment]
> Is there a preference for `enforce=True` to be default, or `False`? I tend to lean towards `True`.

Having `enforce=True` does sound like the most desirable behavior, so I'd be happy to go with that.  In that case, I would still suggest that we change the name of `check_quantity` to something else, since the word "check" implies that no changes are being made and the conversion might be unexpected behavior.  Maybe `validate_quantity` instead?

By the way, thanks for doing this! :cat:

**Comment by StanczakDominik** [comment]
Validate does sound a lot better!

**Comment by rocco8773** [comment]
@namurphy No problem.  I want something like this for [`bapsflib`](https://github.com/BaPSF/bapsflib), so I have selfish motivations. 😁

Well, if we want to maintain backwards compatibility, then the name should remain `check_quantity`.  I also get **"check"** implies that the decorator (by default) should not be changing the arguments, just verifying them. I'd then argue, for an argument to be correct its units must also be equal to the expected units and not just equivalent.  That is, `check_quantity` should raise `ValueError` if the passed argument does not match the expected units.<sup>1</sup>  In this case, I'd change the validation keys up a bit.

1. drop `'enforce'` all together and replace it with `'unit_handling'` (or a better name)
1. `'unit_handling'` would have three values
    1. `'convert'` which would convert the arguments units to the expected units, when possible, and raise a `ValueError` otherwise
    1. `'pass equivalent'` which would allow equivalent units to pass
    1. `None` (i.e. not 1 and 2) which would raise a `ValueError` if the argument units are not equal to the expected units

The downfall of this is...

* `'unit_handling'` is not a simple Boolean
* backwards compatibility is till broken since `{'unit_handling': 'pass equivalent'}` would need to set to maintain current function

<sup>1</sup> Of course, this behavior would automatically break backwards compatibility.

**Comment by rocco8773** [comment]
This could have implications to issue #131 and probably best reflects option 2 pointed out by @namurphy .

>     2. Allow eV to be inputted as temperature, but raise a warning message that a non-standard unit conversion is being made.

In this case, we should have some kind of `UserWarning` that a user could filter out, since repeated know warnings becomes a deterrent to users.  For example, one could do something like...

```
warnings.simplefilter("ignore", catagory=ImplicitUnitConversionWarning)
```

**Comment by StanczakDominik** [comment]
Just a gentle reminder: we're not even at 0.2 yet, so there's little reason to worry about API stability for now - especially since this is something backend-ish for PlasmaPy.

I like the implicit unit conversion warning idea.

**Comment by rocco8773** [comment]
closed by #648 and #722
